### PR TITLE
SOLR-15038: Add elevateDocsWithoutMatchingQ and onlyElevatedReprese…

### PR DESCRIPTION
**Pull request by tkaessmann**
**Same PR like https://github.com/apache/lucene-solr/pull/2134 but this if for Solr 8.**

We've worked a lot with Query Elevation component in the last time and we were missing two features:

Elevate only documents that are part of the search result
In combination with collapsing: Only show the representative if the elevated documents does have the same collapse field value.
Because of this, we've added these two feature toggles elevateDocsWithoutMatchingQ and onlyElevatedRepresentative.

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/handler/component/QueryElevationComponent.java [hunk 1]
```diff
@@ -504,25 +506,27 @@ private void setQuery(ResponseBuilder rb, Elevation elevation) {
...
-      queryBuilder.add(new BoostQuery(elevation.includeQuery, 0f), BooleanClause.Occur.SHOULD);
-      if (elevation.excludeQueries != null) {
-        if (params.getBool(QueryElevationParams.MARK_EXCLUDES, false)) {
-          // We are only going to mark items as excluded, not actually exclude them.
-          // This works with the EditorialMarkerFactory.
-          rb.req.getContext().put(EXCLUDED, elevation.excludedIds);
-        } else {
-          for (TermQuery tq : elevation.excludeQueries) {
-            queryBuilder.add(tq, BooleanClause.Occur.MUST_NOT);
+    if(params.getBool(ELEVATE_DOCS_WITHOUT_MATCHING_Q, true)) {
```
Comment (jbampton) [inline solr/core/src/java/org/apache/solr/handler/component/QueryElevationComponent.java:509]:
```suggestion
    if (params.getBool(ELEVATE_DOCS_WITHOUT_MATCHING_Q, true)) {
```

File: solr/core/src/java/org/apache/solr/search/CollapsingQParserPlugin.java [hunk 2]
```diff
@@ -569,15 +570,18 @@ public int docID() {
...
                              int nullPolicy,
                              IntIntHashMap boostDocsMap,
-                             IndexSearcher searcher) throws IOException {
+                             IndexSearcher searcher,
+                             boolean onlyElevatedRepresentativeVisible) throws IOException {
       this.maxDoc = maxDoc;
       this.contexts = new LeafReaderContext[segments];
+      this.onlyElevatedRepresentativeVisible = onlyElevatedRepresentativeVisible;
       List<LeafReaderContext> con = searcher.getTopReaderContext().leaves();
       for(int i=0; i<con.size(); i++) {
```
Comment (jbampton) [inline solr/core/src/java/org/apache/solr/search/CollapsingQParserPlugin.java:586]:
```suggestion
      for(int i = 0; i < con.size(); i++) {
```

File: solr/core/src/java/org/apache/solr/search/CollapsingQParserPlugin.java [hunk 3]
```diff
@@ -804,9 +812,11 @@ public IntScoreCollector(int maxDoc,
...
                              String field,
                              IntIntHashMap boostDocsMap,
-                             IndexSearcher searcher) {
+                             IndexSearcher searcher,
+                             boolean onlyElevatedRepresentativeVisible) {
       this.maxDoc = maxDoc;
       this.contexts = new LeafReaderContext[segments];
+      this.onlyElevatedRepresentativeVisible = onlyElevatedRepresentativeVisible;
       List<LeafReaderContext> con = searcher.getTopReaderContext().leaves();
       for(int i=0; i<con.size(); i++) {
```
Comment (jbampton) [inline solr/core/src/java/org/apache/solr/search/CollapsingQParserPlugin.java:821]:
```suggestion
      for(int i = 0; i < con.size(); i++) {
```

File: solr/core/src/java/org/apache/solr/search/CollapsingQParserPlugin.java [hunk 4]
```diff
@@ -1028,25 +1043,25 @@ public OrdFieldValueCollector(int maxDoc,
       this.needsScores4Collapsing = needsScores4Collapsing;
       this.needsScores = needsScores;
       if (null != sortSpec) {
-        this.collapseStrategy = new OrdSortSpecStrategy(maxDoc, nullPolicy, valueCount, groupHeadSelector, this.needsScores4Collapsing, this.needsScores, boostDocs, sortSpec, searcher, collapseValues);
+        this.collapseStrategy = new OrdSortSpecStrategy(maxDoc, nullPolicy, valueCount, groupHeadSelector, this.needsScores4Collapsing, this.needsScores, boostDocs, sortSpec, searcher, onlyElevatedRepresentativeVisible);
       } else if (funcQuery != null) {
-        this.collapseStrategy =  new OrdValueSourceStrategy(maxDoc, nullPolicy, valueCount, groupHeadSelector, this.needsScores4Collapsing, this.needsScores, boostDocs, funcQuery, searcher, collapseValues);
+        this.collapseStrategy =  new OrdValueSourceStrategy(maxDoc, nullPolicy, valueCount, groupHeadSelector, this.needsScores4Collapsing, this.needsScores, boostDocs, funcQuery, searcher, onlyElevatedRepresentativeVisible);
```
Comment (jbampton) [inline solr/core/src/java/org/apache/solr/search/CollapsingQParserPlugin.java:1048]:
```suggestion
        this.collapseStrategy = new OrdValueSourceStrategy(maxDoc, nullPolicy, valueCount, groupHeadSelector, this.needsScores4Collapsing, this.needsScores, boostDocs, funcQuery, searcher, onlyElevatedRepresentativeVisible);
```

File: solr/core/src/java/org/apache/solr/search/CollapsingQParserPlugin.java [hunk 5]
```diff
@@ -1028,25 +1043,25 @@ public OrdFieldValueCollector(int maxDoc,
...
             break;
           }
           case FLOAT: {
-            this.collapseStrategy = new OrdFloatStrategy(maxDoc, nullPolicy, valueCount, groupHeadSelector, this.needsScores, boostDocs, collapseValues);
+            this.collapseStrategy = new OrdFloatStrategy(maxDoc, nullPolicy, valueCount, groupHeadSelector, this.needsScores, boostDocs, onlyElevatedRepresentativeVisible);
             break;
           }
           case LONG: {
-            this.collapseStrategy =  new OrdLongStrategy(maxDoc, nullPolicy, valueCount, groupHeadSelector, this.needsScores, boostDocs, collapseValues);
+            this.collapseStrategy =  new OrdLongStrategy(maxDoc, nullPolicy, valueCount, groupHeadSelector, this.needsScores, boostDocs, onlyElevatedRepresentativeVisible);
```
Comment (jbampton) [inline solr/core/src/java/org/apache/solr/search/CollapsingQParserPlugin.java:1064]:
```suggestion
            this.collapseStrategy = new OrdLongStrategy(maxDoc, nullPolicy, valueCount, groupHeadSelector, this.needsScores, boostDocs, onlyElevatedRepresentativeVisible);
```

File: solr/solrj/src/java/org/apache/solr/common/params/QueryElevationParams.java [hunk 6]
```diff
@@ -55,4 +55,16 @@
...
+   * If you only want to elevate the docs that are part of the search result, set this to false. True by default.
+   */
+  String ELEVATE_DOCS_WITHOUT_MATCHING_Q = "elevateDocsWithoutMatchingQ";
+
+  /**
+   * If elevation is used in combination with the collapse query parser, we can define that we only want to return the
+   * representative and not all elevated docs.
+   */
+  String ONLY_ELEVATED_REPRESENTATIVE = "onlyElevatedRepresentative";
 }
```
Comment (jbampton) [inline solr/solrj/src/java/org/apache/solr/common/params/QueryElevationParams.java:70]:
```suggestion
}

```

**Comment by tkaessmann** [comment]
not needed anymore

**Pull request closed without merging**